packages/core/rfcs/20200401-products-filters.md

```diff
@@ -0,0 +1,81 @@
+# Filtering products
+
+Filtering products is an important part of any products listing page. To ensure the best developer experience and consistency between integrations (because filters have to be delivered to the theme in a unified form) we need to provide a common way to retrieve available filters and unified interface for them.
+
+## New APIs
+
+### Modify `useProduct` composable
+
+Since there are several factors that the filters depend on (products type, category, available products, their attributes) and the set of filters might be dynamic, this feature should introduce new fields in the `useProduct` composable, those are `availableFilters` and `chosenFilters`, like in the example below.
+
+**New interface**:
+
+```typescript
+export interface UseProduct<PRODUCT, FILTER_BASE> {
+  products: ComputedProperty<PRODUCT[]>;
+  totalProducts: ComputedProperty<number>;
+  availableFilters: ComputedProperty<Filter<FILTER_BASE>[]>;
+  selectedFilters: ComputedProperty<Filter<FILTER_BASE>[]>
+  search: (params: {
+    perPage?: number;
+    page?: number;
+    sort?: any;
+    term?: any;
+    filters?: Filter<FILTER_BASE>[];
+    [x: string]: any;
+  }) => Promise<void>;
+  loading: ComputedProperty<boolean>;
+  [x: string]: any;
+}
+```
+
+**`UseProductFactoryParams` interface**
+
+Loading filters, their possible values, counts, etc. will require an integration-specific logic that will be part of the search function, so the factory params interface stays intact, but we need a new interface for search function result:
+
+```typescript
+export interface FilteredSearchResult<T, FILTER_BASE> extends SearchResult<T> {
+  availableFilters: Filter<FILTER_BASE>[];
+  selectedFilters: Filter<FILTER_BASE>[];
+}
+```
+
+### Filter interface
+
+We need a unified interface for filters, like:
+
+```typescript
+// FILTER_BASE is an enum with available filter bases, like price, attribute, status, quantity, availability, sale percentage etc.
+// this will be handled per integration, but basic agnostic bases are added
+interface Filter<FILTER_BASE> {
+  name: string;
+  type: FilterType;
+  filterBase: AgnosticFilterBase | FILTER_BASE;
+  options: FilterOption[]
+}
+
+enum FilterType {
+  RANGE = 'range', // eg. price
```

Comment:
i dont think every platform is providing something like that, its helpful but maybe too ui-specific?

Reply:
I don't think it is ui-specific. Such filter type not only allows ui to choose proper component or its settings, but also provides a hint for underlying logic - how should we map filter options to something that api client will handle in request. 
So `range` will determine `min < x < max`, `single` will only allow single option to be selected, `multi` will allow multiple options to be selected and apply logical alternative on them.

Reply:
yup, but its still additional work for the integrators, the question is - does it provide enough value to justify this additional work? In my opinion it's not, mostly because you can have multiple filterable fields and all of them are depending on platform and sometimes on settings from the admin panel. As a rule of thumb we shouldn't add anything that is not coming from the platform sites;f. I would see something like that more as an API extension than something mandatory for the integration